lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1435,6 +1456,1088 @@ class BuildIfGroups : public calyx::FuncOpPartialLoweringPattern {
+      for (Block *blk : iset) {
+        if (blk == memUserBlock)
+          continue; // Skip the block containing memUser
+        if (eligibleOpsCache.find(blk) == eligibleOpsCache.end()) {
+          std::vector<Operation *> ops;
+          for (auto &op : blk->getOperations()) {
+            if (potentialOps.find(&op) != potentialOps.end()) {
+              ops.push_back(&op);
+              break;
+            }
+          }
+          eligibleOpsCache[blk] = std::move(ops);
+        }
+      }
+    }
+    // Step 2: Iterate over each maximal independent set
+    for (const auto &iset : maximalISets) {
+      SmallVector<Operation *, 4> currentComb;
+      currentComb.push_back(memUser);
+      std::vector<std::vector<Operation *>> eligibleOpsPerBlock;
+      eligibleOpsPerBlock.reserve(iset.size());
+      bool skipSet = false;
+      for (Block *blk : iset) {
+        if (blk == memUserBlock)
+          continue;
+        auto it = eligibleOpsCache.find(blk);
+        if (it == eligibleOpsCache.end() || it->second.empty()) {
+          skipSet = true;
+          break;
+        }
+        eligibleOpsPerBlock.push_back(it->second);
+      }
+      if (skipSet)
+        continue;
+      // Iterative backtracking
+      size_t numBlocks = eligibleOpsPerBlock.size();
+      SmallVector<size_t, 4> indices(numBlocks, 0);
+      while (true) {
+        // Build the current combination
+        std::vector<Operation *> comb;
+        comb.push_back(memUser);
+        for (size_t i = 0; i < numBlocks; ++i) {
+          comb.push_back(eligibleOpsPerBlock[i][indices[i]]);
+        }
+        allCombs.emplace_back(std::move(comb));
+        size_t block = numBlocks;
+        while (block > 0) {
+          block--;
+          if (indices[block] + 1 < eligibleOpsPerBlock[block].size()) {
+            indices[block]++;
+            // Reset all subsequent indices
+            for (size_t j = block + 1; j < numBlocks; ++j) {
+              indices[j] = 0;
+            }
+            break;
+          }
+        }
+        if (block == 0 && indices[0] + 1 >= eligibleOpsPerBlock[0].size())
+          break; // All combinations are generated
+      }
+    }
+  }
+  // TODO: maybe we can turn `Trace` into a class
+  SmallVector<SmallVector<Value>>
+  initCompress(SmallVector<SmallVector<Value>> &unCompressedTraces) const {
+    // 1. Remove redundant info: memory access with the same address in the same
+    // step
+    SmallVector<SmallVector<Value>> redundantRemoved;
+    for (const auto &step : unCompressedTraces) {
+      DenseSet<int> seenAddr;
+      SmallVector<Value> simplifiedStep;
+      for (const auto addrVal : step) {
+        auto addrInt = evaluateIndex(addrVal);
+        if (seenAddr.insert(addrInt).second)
+          simplifiedStep.push_back(addrVal);
+      }
+      redundantRemoved.push_back(simplifiedStep);
+    }
+    // 2. Combine steps with identical access into a single step
+    std::set<std::set<int>> seenStep;
```

Comment:
similarly here, I couldn't get a `DenseSet` of `DenseSet` working unless I implement extra methods, so not entirely sure what to do.